plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java

```diff
@@ -463,21 +463,40 @@ public static ColumnMapping timestampColumnMappingUsingSqlTimestampWithRounding(
         checkArgument(timestampType.getPrecision() <= TimestampType.MAX_SHORT_PRECISION, "Precision is out of range: %s", timestampType.getPrecision());
         return ColumnMapping.longMapping(
                 timestampType,
-                (resultSet, columnIndex) -> {
-                    LocalDateTime localDateTime = resultSet.getTimestamp(columnIndex).toLocalDateTime();
-                    int roundedNanos = toIntExact(round(localDateTime.getNano(), 9 - timestampType.getPrecision()));
-                    LocalDateTime rounded = localDateTime
-                            .withNano(0)
-                            .plusNanos(roundedNanos);
-                    return toTrinoTimestamp(timestampType, rounded);
-                },
+                timestampWithRoundingReadFunction(timestampType),
                 timestampWriteFunctionUsingSqlTimestamp(timestampType),
                 // NOTE: pushdown is disabled because the values stored in remote system might not match the values as
                 // read by Trino due to rounding. This can lead to incorrect results if operations are pushed down to
                 // the remote system.
                 DISABLE_PUSHDOWN);
     }
 
+    @Deprecated
+    public static ColumnMapping timestampColumnMappingUsingSqlTimestampWithRoundingFullPushdown(TimestampType timestampType)
+    {
+        // TODO support higher precision
+        checkArgument(timestampType.getPrecision() <= TimestampType.MAX_SHORT_PRECISION, "Precision is out of range: %s", timestampType.getPrecision());
+        return ColumnMapping.longMapping(
+                timestampType,
+                timestampWithRoundingReadFunction(timestampType),
+                timestampWriteFunctionUsingSqlTimestamp(timestampType),
+                // NOTE: as noted in the "DISABLE_PUSHDOWN" version, the values stored in remote system might not match
+                // the values as read by Trino due to rounding.
+                FULL_PUSHDOWN);
```

Comment:
@jerryleooo Won't FULL_PUSHDOWN with rounding cause correctness issues?

Reply:
@jibinpt so far I haven't found correctness issue

Reply:
There can be issues where if you send a query like `WHERE _time < TIMESTAMP '2020-01-01 01:01:01.999111'`. The rounding will actually send the query as `_time < TIMESTAMP '2020-01-01 01:01:01.999'` meaning that the result will have missing rows if there are any with `TIMESTAMP '2020-01-01 01:01:01.999' < _time < TIMESTAMP '2020-01-01 01:01:01.999111'`.

The tests will uncover this once you copy all the cases from the PostgreSQL (or SQL Server) type mapping tests.

Reply:
Hi @hashhar 
1. In Druid the smallest time unit is millisecond: https://druid.apache.org/docs/latest/querying/granularities.html
2. I think I can not copy the cases directly since Druid only supports 5 data types: https://druid.apache.org/docs/latest/querying/sql.html#standard-types

Reply:
Oh, I don't mean to do it all at once. I just meant to copy some of the values that we test for timestamps in the PostgreSQL type mapping tests. Sorry for being unclear.

Reply:
@jerryleooo @hashhar if smallest time unit supported by Druid is milliseconds, then there shouldn't be any correctness issue with the current logic right? Are we concerned that if in future Druid supports more precision(like micro seconds), there could be correctness issues down the line? Is there any issue if we create and use an equivalent function without any rounding logic similar to below?

public static ColumnMapping timestampColumnMappingUsingSqlTimestamp(TimestampType timestampType)
    {
        // TODO support higher precision
        checkArgument(timestampType.getPrecision() <= TimestampType.MAX_SHORT_PRECISION, "Precision is out of range: %s", timestampType.getPrecision());
        return ColumnMapping.longMapping(
                timestampType,
                (resultSet, columnIndex) -> {
                    LocalDateTime localDateTime = resultSet.getTimestamp(columnIndex).toLocalDateTime();
                    return toTrinoTimestamp(timestampType, localDateTime);
                },
                timestampWriteFunctionUsingSqlTimestamp(timestampType),
                FULL_PUSHDOWN);
    }

Reply:
Continued at https://github.com/trinodb/trino/pull/8474#discussion_r670985264.